File: RecyclerViewProduct/app/src/main/java/com/example/recyclerviewproduct/MainActivity.java
package com.example.recyclerviewproduct;

import android.os.Bundle;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {
    private RecyclerView productsRecyclerView;
    private ProductsAdapter productsAdapter;
    private ArrayList<Product> productArrayList;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initDataSet();
        productsRecyclerView = findViewById(R.id.productsRecyclerView);
        productsAdapter = new ProductsAdapter(productArrayList);
        productsRecyclerView.setAdapter(productsAdapter);
        productsRecyclerView.setLayoutManager(new LinearLayoutManager(this,
                LinearLayoutManager.VERTICAL,
                false));
    }

    private void initDataSet(){
        productArrayList = new ArrayList<Product>();
        for(int i =0;i<10;i++) {
            productArrayList.add(new Product(i + 1,
                    "Product " + i,
                    i * 5 + 10,
                    i
            ));
        }
    }
}

File: Webservices/app/src/main/java/com/example/webservices/MainActivity.kt
package com.example.webservices

import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.os.StrictMode
import android.util.Log
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import org.json.JSONArray
import org.json.JSONObject
import java.net.URL
import javax.net.ssl.HttpsURLConnection

class MainActivity : AppCompatActivity() {

    private lateinit var recyclerViewForUsers: RecyclerView
    private var users = ArrayList<User>()
    private lateinit var usersAdapter : UsersAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        //        StrictMode.setVmPolicy(StrictMode.VmPolicy.Builder().build());
        var policy = StrictMode.ThreadPolicy.Builder().permitAll().build()
        StrictMode.setThreadPolicy(policy)
        WebThread(MessageHandler()).execute()
        initRecyclerView()
    }
    private fun initRecyclerView(){
        recyclerViewForUsers = findViewById(R.id.recyclerViewForUsers)
        usersAdapter = UsersAdapter(users)
        recyclerViewForUsers.adapter = usersAdapter
        recyclerViewForUsers.layoutManager = LinearLayoutManager(this,LinearLayoutManager.VERTICAL,false)
    }
    inner class MessageHandler : Handler(){
        override fun handleMessage(msg: Message) {
            super.handleMessage(msg)
            var usersResponse = msg.obj as ArrayList<User>

            //refer to users array from outer class i.e. MainActivity
            this@MainActivity.users.addAll(usersResponse)
            usersAdapter.notifyDataSetChanged()   //imp
        }
    }
}